### PR TITLE
Rewrite DNS code to get other record types

### PR DESCRIPTION
**Pull request by erikcorry**

### CODE REVIEW

File: lib/net/modules/dns.toit [hunk 1]
```diff
@@ -356,7 +455,8 @@ parts_ packet/ByteArray position/int parts/List [position-block] -> none:
   position-block.call position + 1
 
 /// Limits the size of the cache to avoid using too much memory.
-trim-cache_ cache/Map -> none:
+trim-cache_ map/Map type/int -> none:
```
Comment (kasperl) [inline lib/net/modules/dns.toit:458]: Could this be static?

File: lib/net/modules/dns.toit [hunk 2]
```diff
@@ -303,31 +380,53 @@ class DnsClient:
...
+          len := type == RECORD-A ? 4 : 16
+          if rd-length != len: throw (DnsException "Unexpected IP address length $rd-length")
+          result.add
+              net.IpAddress
+                  response.copy position position + len
+        else if type == RECORD-CNAME or type == RECORD-PTR:
+          result.add
+              decode-name response position: null
+        else if type == RECORD-TXT:
+          if response[position] != rd-length - 1: throw "WTF"
```
Comment (kasperl) [inline lib/net/modules/dns.toit:398]: Maybe improve the thrown exception here.

File: lib/net/modules/dns.toit [hunk 3]
```diff
@@ -249,42 +319,49 @@ class DnsClient:
...
-      if cache_.contains name:
-        entry := cache_[name]
-        if entry.valid: return entry.address
-        cache_.remove name
-    if accept-ipv6:
-      if cache-ipv6_.contains name:
-        entry := cache-ipv6_[name]
-        if entry.valid: return entry.address
-        cache-ipv6_.remove name
+  find-in-cache_ name type/int -> List?:
```
Comment (kasperl) [inline lib/net/modules/dns.toit:322]: static?

File: lib/net/modules/dns.toit [hunk 4]
```diff
@@ -249,42 +319,49 @@ class DnsClient:
...
+    if cache_.contains type:
+      map := cache_[type]
+      if map.contains name:
+        entry := map[name]
+        if entry.valid:
+          return entry.value
+        else:
+          map.remove name
```
Comment (kasperl) [inline lib/net/modules/dns.toit:330]: Not sure it is much better, but:

```suggestion
    map := cache_.get type --if-absent=: return null
    entry := map.get name --if-absent=: return null
    if entry.valid: return entry.value
    map.remove name
    return null
```

File: lib/net/modules/dns.toit [hunk 5]
```diff
@@ -303,31 +380,53 @@ class DnsClient:
...
-        if case-compare_ r-name q-name:
-          result := net.IpAddress
-              response.copy position position + 4
-          if ttl > 0:
-            trim-cache_ cache_
-            cache_[query.name] = CacheEntry_ result ttl
-          return result
+      if type == overall-type and case-compare_ r-name q-name:
+        if type == RECORD-A or type == RECORD-AAAA:
+          len := type == RECORD-A ? 4 : 16
```
Comment (kasperl) [inline lib/net/modules/dns.toit:389]: You use `length` in other places.

File: lib/net/modules/dns.toit [hunk 6]
```diff
@@ -303,31 +380,53 @@ class DnsClient:
...
+          if rd-length != len: throw (DnsException "Unexpected IP address length $rd-length")
+          result.add
+              net.IpAddress
+                  response.copy position position + len
+        else if type == RECORD-CNAME or type == RECORD-PTR:
+          result.add
+              decode-name response position: null
+        else if type == RECORD-TXT:
+          if response[position] != rd-length - 1: throw "WTF"
+          len := min (rd-length - 1) response[position]
```
Comment (kasperl) [inline lib/net/modules/dns.toit:399]: You use `length` in other places.

File: lib/net/modules/dns.toit [hunk 7]
```diff
@@ -303,31 +380,53 @@ class DnsClient:
...
         q-name = decode-name response position: null
       position += rd-length
-    throw (DnsException "Response did not contain matching A record" --name=query.name)
+    if result.size > 0:
+      if min-ttl > 0:
+        trim-cache_ cache_ overall-type
+        type-cache := cache_.get overall-type --init=: {:}
+        type-cache[query.name] = CacheEntry_ result min-ttl
+      return result
+    return []
```
Comment (kasperl) [inline lib/net/modules/dns.toit:421]:
```suggestion
    if result.size > 0 and min-ttl > 0:
      trim-cache_ cache_ overall-type
      type-cache := cache_.get overall-type --init=: {:}
      type-cache[query.name] = CacheEntry_ result min-ttl
    return result
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: lib/net/modules/dns.toit (hunk 1 above)
Comment (erikcorry) [reply lib/net/modules/dns.toit:480]: I moved it into the class.

File: lib/net/modules/dns.toit (hunk 2 above)
Comment (erikcorry) [reply lib/net/modules/dns.toit:420]: Done.

File: lib/net/modules/dns.toit (hunk 3 above)
Comment (erikcorry) [reply lib/net/modules/dns.toit:344]: Uses the cache, which is a field.  Same applies to trim-cache_.

File: lib/net/modules/dns.toit [hunk 8]
```diff
@@ -39,7 +60,7 @@ If there are multiple servers then they are tried in rotation until one
   the next one.  This is in line with the way that Linux handles multiple
   servers on the same lookup request.
 */
-dns-lookup -> net.IpAddress
+dns-lookup-ips -> List
```
Comment (kasperl) [inline lib/net/modules/dns.toit:63]: Maybe `-all` or `-multi` instead of `-ips`?

Comment (erikcorry) [reply lib/net/modules/dns.toit:63]: I like -multi

**Pull request merged**